llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -10526,6 +10527,7 @@ InstructionCost BoUpSLP::getGatherCost(ArrayRef<Value *> VL,
     if (!UniqueElements.insert(V).second) {
       DuplicateNonConst = true;
       ShuffledElements.setBit(I);
+      ShuffleMask[I] = I;
```

Comment:
This does not look correct. You generated identity mask here, which is not. Need to record the position of the first unique value and then use it here as index.

Reply:
Clearly it is time for me to stop for the day.  I wrote the correct code here - and then managed to commit the wrong version.  Oops.  Will push a corrected version tomorrow.